pkg/controllers/management/clusterupstreamrefresher/cluster_upstream_refresher.go

```diff
@@ -0,0 +1,323 @@
+
+	eksv1 "github.com/rancher/eks-operator/pkg/apis/eks.cattle.io/v1"
+	apimgmtv3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	v3 "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	mgmtv3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/rancher/rancher/pkg/wrangler"
+	wranglerv1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/robfig/cron"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	noKEv2Provider           = "none"
+	clusterLastRefreshTime   = "clusters.management.cattle.io/ke-last-refresh"
+	refreshCronSettingFormat = "%s-refresh"
+	eksRefreshCronDeprecated = "eks-refresh-cron"
+	eksRefreshCronMigrated   = "clusters.management.cattle.io/is-eks-cron-migrated"
+)
+
+type clusterRefreshController struct {
+	secretsCache        wranglerv1.SecretCache
+	settingsClient      v3.SettingClient
+	settingsCache       v3.SettingCache
+	clusterClient       v3.ClusterClient
+	clusterCache        v3.ClusterCache
+	clusterEnqueueAfter func(name string, duration time.Duration)
+}
+
+// for other cloud drivers, please edit HERE
+type clusterConfig struct {
+	eksConfig *eksv1.EKSClusterConfigSpec
+}
+
+func Register(ctx context.Context, wContext *wrangler.Context) {
+	c := clusterRefreshController{
+		secretsCache:        wContext.Core.Secret().Cache(),
+		settingsClient:      wContext.Mgmt.Setting(),
+		settingsCache:       wContext.Mgmt.Setting().Cache(),
+		clusterClient:       wContext.Mgmt.Cluster(),
+		clusterCache:        wContext.Mgmt.Cluster().Cache(),
+		clusterEnqueueAfter: wContext.Mgmt.Cluster().EnqueueAfter,
+	}
+
+	wContext.Mgmt.Cluster().OnChange(ctx, "cluster-refresher-controller", c.onClusterChange)
+}
+
+func (c *clusterRefreshController) onClusterChange(key string, cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	if cluster == nil || cluster.DeletionTimestamp != nil {
+		return cluster, nil
+	}
+
+	provider, ready := getProviderAndReadyStatus(cluster)
+	if provider == noKEv2Provider {
+		// not a KEv2 cluster
+		return cluster, nil
+	}
+
+	if !ready {
+		logrus.Infof("initial upstream spec for cluster [%s] has not been set by cluster handler yet, skipping", cluster.Name)
+		return cluster, nil
+	}
+
+	var lastRefreshTime string
+	var err error
+
+	if cluster.Annotations != nil {
+		lastRefreshTime = cluster.Annotations[clusterLastRefreshTime]
+	}
+
+	// migrate deprecated EKS refresh cron setting to new format
+	if provider == apimgmtv3.ClusterDriverEKS &&
+		(cluster.Annotations == nil || cluster.Annotations[eksRefreshCronMigrated] == "") {
+		cluster, err = c.migrateEksRefreshCronSetting(cluster)
+		if err != nil {
+			return cluster, err
+		}
+		cluster.Annotations[eksRefreshCronMigrated] = "true"
+		if cluster, err = c.clusterClient.Update(cluster); err != nil {
+			return cluster, err
+		}
```

Comment:
It feels bad but I'm sure it' not an expensive call. Annotation seems heavy too. Can we really not remove settings? If we could only remove deprecated settings none of this would have to be weird.

Reply:
Either way it wouldn't make sense to put the annotation on the cluster since this needs to be performed one time, not once for each cluster.

Reply:
As an update, the annotation will now go away entirely with our new plan.

Reply:
I agree, annotation is not needed. I have re-implemented this once again